### PR TITLE
feat: Support subscript and nested functions in grouping queries 

### LINKED ISSUE #5906: Subscript Expressions in GROUP BY not valid as non-aggregate SELECT function arguments
**Describe the bug**

When aggregating a stream with a `GROUP BY` expression that involves a subscript expression (i.e. map lookup, `col1['foo']` the AggregateAnalyzer will not properly accept them as function arguments to non-aggregate functions.

**To Reproduce**

```json
    {
      "name": "Subscript Expressions in GROUP BY not valid non-aggregate SELECT function arguments",
      "statements": [
        "CREATE STREAM input (id INT KEY, col1 MAP<VARCHAR, VARCHAR>) WITH (kafka_topic='input', value_format='json');",
        "CREATE TABLE output AS SELECT col1['foo'], AS_VALUE(col1['foo']) AS foo, COUNT(*) FROM input WINDOW TUMBLING (SIZE 10 MILLISECONDS) GROUP BY col1['foo'];"
      ],
      "inputs": [],
      "outputs": []
    },
```

**Expected behavior**

The query succeeds.

**Actual behaviour**

```
Caused by: io.confluent.ksql.util.KsqlException: Non-aggregate SELECT expression(s) not part of GROUP BY: AS_VALUE(COL1['foo'])
Either add the column to the GROUP BY or remove it from the SELECT.
	at io.confluent.ksql.analyzer.AggregateAnalyzer$AggAnalyzer.enforceAggregateRules(AggregateAnalyzer.java:258)
	at io.confluent.ksql.analyzer.AggregateAnalyzer$AggAnalyzer.result(AggregateAnalyzer.java:240)
	at io.confluent.ksql.analyzer.AggregateAnalyzer.analyze(AggregateAnalyzer.java:62)
	at io.confluent.ksql.planner.LogicalPlanner.buildAggregateNode(LogicalPlanner.java:234)
	at io.confluent.ksql.planner.LogicalPlanner.buildPlan(LogicalPlanner.java:132)
	at io.confluent.ksql.engine.QueryEngine.buildQueryLogicalPlan(QueryEngine.java:74)
	at io.confluent.ksql.engine.EngineExecutor.planQuery(EngineExecutor.java:202)
	at io.confluent.ksql.engine.EngineExecutor.plan(EngineExecutor.java:161)
	at io.confluent.ksql.engine.KsqlEngine.plan(KsqlEngine.java:186)
	at io.confluent.ksql.test.tools.TestExecutorUtil$PlannedStatementIterator.planStatement(TestExecutorUtil.java:422)
	... 40 more
```

**Additional context**

Note that the same expression will work if the group by column is not a subscript expression, such as:
```json
      "statements": [
        "CREATE STREAM input (id INT KEY, col1 VARCHAR) WITH (kafka_topic='input', value_format='json');",
        "CREATE TABLE output AS SELECT col1, AS_VALUE(col1) AS foo, COUNT(*) FROM input WINDOW TUMBLING (SIZE 10 MILLISECONDS) GROUP BY col1;"
      ],
```

### PR DESCRIPTION
**Pull request by vpapavas**
### Description 
Fixes #5906 and #5967 

### Testing done 
Added test cases in `group-by.json`

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java [hunk 1]
```diff
@@ -204,9 +226,18 @@ private void throwOnWindowBoundColumnIfWindowedAggregate(
...
-      final boolean matchesGroupBy = groupBy.contains(expression);
+
+      boolean matchesGroupBy = false;
+      // If the non-Agg expression is a function, then all its arguments must be part of the
+      // grouping columns. Note, they may be nested inside other functions.
+      /*if (expression instanceof FunctionCall) {
+        matchesGroupBy = functionContainsOnlyGroupingColumns(expression, true);
+      } else {
+        matchesGroupBy = groupBy.contains(expression);
+      }*/
```
Comment (agavra) [inline ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:239]: is there a reason this code is still commented out?

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java [hunk 2]
```diff
@@ -221,15 +252,15 @@ private void captureNonAggregateSelectNotPartOfGroupBy(
...
+        final Set<Expression> nonAggParams
     ) {
       nonAggParams.stream()
           .filter(param -> !groupBy.contains(param))
           .forEach(aggSelectsNotPartOfGroupBy::add);
     }
 
-    private void captureNoneAggregateHavingNotPartOfGroupBy(final ColumnReferenceExp nonAggColumn) {
-      if (groupBy.contains(new UnqualifiedColumnReferenceExp(nonAggColumn.getColumnName()))) {
+    private void captureNoneAggregateHavingNotPartOfGroupBy(final Expression nonAggColumn) {
```
Comment (agavra) [inline ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:262]:
```suggestion
    private void captureNonAggregateHavingNotPartOfGroupBy(final Expression nonAggColumn) {
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java [hunk 3]
```diff
@@ -344,8 +384,10 @@ public Void visitUnqualifiedColumnReference(
         final UnqualifiedColumnReferenceExp node,
         final Void context
     ) {
-      dereferenceCollector.accept(aggFunctionName, node);
-
+      if (!foundExpressionInGroupBy || visitedAggFunction) {
+        dereferenceCollector.accept(aggFunctionName, node);
+      }
+      foundExpressionInGroupBy = false;
```
Comment (agavra) [inline ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:390]: I'm a little confused here, why do we unset the `foundExpressioninGroupBy` here? I feel like we should set it right after we set it and call `super`. For example we should unset it after calling `super.visitSubscriptExpression(node, context);`

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java [hunk 4]
```diff
@@ -359,5 +401,14 @@ public Void visitQualifiedColumnReference(
     ) {
       throw new UnsupportedOperationException("Should of been converted to unqualified");
     }
+
+    @Override
+    public Void visitSubscriptExpression(final SubscriptExpression node, final Void context) {
```
Comment (agavra) [inline ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:406]: are there any other expressions that we should be looking that can match exactly other than function calls and subscript expressions? i.e. what about arithmetic expressions (`SELECT AS_VALUE(a + b) AS foo GROUP BY a + b`)

I think it makes sense to implement this in `visitExpression` instead of in each individual expression type. We would only "override" this behavior when we need to (e.g. inside `visitFuncitonCall`)

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json [hunk 5]
```diff
@@ -511,6 +605,17 @@
         "message": "Duplicate value columns found in schema: `COUNT` BIGINT"
       }
     },
+    {
+      "name": "map used in non-aggregate function in select when group by uses subscript",
```
Comment (agavra) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json:609]: one more test (may or may not already be there) - we should group by a map and select a value within the map as a column:
```sql
SELECT col1, col1['foo'] AS foo FROM .. GROUP BY col1;
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java [hunk 6]
```diff
@@ -68,17 +69,19 @@ public AggregateAnalysisResult analyze(
     private final MutableAggregateAnalysis aggregateAnalysis = new MutableAggregateAnalysis();
     private final FunctionRegistry functionRegistry;
     private final Set<Expression> groupBy;
+    private boolean foundExpressionInGroupBy = false;
```
Comment (agavra) [inline ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:72]: why can't this be part of the `Visitor` directly? it seems weird for this to carry over between process steps (lines 95 -> 109)

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json [hunk 7]
```diff
@@ -362,6 +362,100 @@
         ]
       }
     },
+    {
+      "name": "subscript in group-by and select",
```
Comment (agavra) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json:366]: can we add corresponding tests for `HAVING`?

**Comment by agavra** [comment]
Also don't forget to check in the historical plans:
```
[2020-08-13T01:41:47.476Z] [ERROR]   PlannedTestsUpToDateTest.shouldHaveLatestPlans:87 Current query plan differs from latest for: group-by - subscript in group-by and select. 
18:41:47  location: file:///home/jenkins/workspace/entinc_Contributors_ksql_PR-5998/ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json:366
18:41:47  Please re-generate QTT plans by running temporarily commenting out the @Ignore andthen running PlannedTestGeneratorTest.manuallyGeneratePlans().
18:41:47  See `ksqldb-functional-tests/README.md` for more info.
18:41:47  Expected: no saved plan
18:41:47       but: was <io.confluent.ksql.test.planned.TestCasePlan@3b7d8d2e>
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java (hunk 1 above)
Comment (vpapavas) [reply ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:239]: Sorry, forgot to remove

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java (hunk 4 above)
Comment (vpapavas) [reply ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:406]: Yes, this is a good point in the sense that we need to figure out which expressions need special handling, I will try the arithmetic expressions. However, there is no `visitExpression` in `TraversalExpressionVisitor` unless I misunderstand what you meant.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json (hunk 7 above)
Comment (vpapavas) [reply ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json:366]: I am a bit weirded out but the behavior of `HAVING`: Why does it return rows for non-matching keys with null value?

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json (hunk 5 above)
Comment (vpapavas) [reply ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json:609]: Grouping by a map throws the exception: 
```
Stacktrace was: io.confluent.ksql.SchemaNotSupportedException: Key format does not support key schema.
format: KAFKA
schema: Persistence{schema=STRUCT<COL1 MAP<VARCHAR, VARCHAR>> NOT NULL, unwrapped=false}
reason: The 'KAFKA' format does not support type 'MAP'
```

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json (hunk 7 above)
Comment (big-andy-coates) [reply ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json:366]: Because of this: https://github.com/confluentinc/ksql/issues/3558

**Comment by big-andy-coates** [comment]
Look at `PlannedTestGeneratorTest` for generating the plans.

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java (hunk 6 above)
Comment (big-andy-coates) [reply ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:72]: +1

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java [hunk 8]
```diff
@@ -128,48 +137,61 @@ private void processSelect(final Expression expression) {
...
       // Window bounds are supported for operations on windowed sources
       if (!analysis.getWindowExpression().isPresent()) {
         return;
       }
 
       // For non-windowed sources, with a windowed GROUP BY, they are only supported in selects:
-      if (SystemColumns.isWindowBound(node.getColumnName())) {
+      if (!(node instanceof ColumnReferenceExp)) {
+        return;
+      }
```
Comment (big-andy-coates) [inline ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:193]: The comment on line 190 refers to code now on line 194.  Can you move this `if` statement above the comment please?

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java (hunk 1 above)
Comment (big-andy-coates) [reply ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:239]: +1

**Comment by big-andy-coates** [comment]
Hey @vpapavas,

There's another potential issue around this you may have fixed, or which is worth looking at at the same time.

The thread is in community slack: https://confluentcommunity.slack.com/archives/C6UJNMY67/p1594325261431500

However, the short version is that this doesn't currently work:

```sql
CREATE TABLE my_agg_z_t AS
    SELECT   
    id as id_key,
    number as number_key,
    CAST(id as STRING)+'|+|'+CAST(number as STRING) AS crk,
    AS_VALUE(CAST(id as STRING)) as id,
    AS_VALUE(CAST(number as STRING)) as number,
    AVG(value) AS someavg FROM blah_s
    WINDOW TUMBLING (size 1 hour)
    GROUP BY CAST(id as STRING)+'|+|'+CAST(number as STRING)
    EMIT CHANGES;
```

Resulting in the error:

```
Non-aggregate SELECT expression(s) not part of GROUP BY: ID, NUMBER, AS_VALUE(CAST(ID AS STRING)), AS_VALUE(CAST(NUMBER AS STRING))
Either add the column to the GROUP BY or remove it from the SELECT.
```

I think this PR fixes _some_ of this, but maybe not all.  It would be good to add a test case that copies the key columns, i.e. the grouping expressions, into the value using `AS_VALUE`, where the grouping expressions are not simple columns names,  as is being done above.

**Comment by vpapavas** [comment]
Hey @big-andy-coates ,

the query you posted shouldn't work, i.e., it is correct that it fails. One can only use grouping columns in non-aggregate functions. Namely, the grouping column in this example is `CAST(id as STRING)+'|+|'+CAST(number as STRING)`. So, only this entire expression can appear in the `AS_VALUE` function like so `AS_VALUE(CAST(id as STRING)+'|+|'+CAST(number as STRING))`.

For the query to work as is one workaround is for the grouping columns to contain additionally `id` and `number` like so:
```
CREATE TABLE my_agg_z_t AS
    SELECT   
    id as id_key,
    number as number_key,
    CAST(id as STRING)+'|+|'+CAST(number as STRING) AS crk,
    AS_VALUE(CAST(id as STRING)) as id,
    AS_VALUE(CAST(number as STRING)) as number,
    AVG(value) AS someavg FROM blah_s
    WINDOW TUMBLING (size 1 hour)
    GROUP BY col1, col2, CAST(id as STRING)+'|+|'+CAST(number as STRING)
    EMIT CHANGES;
```
and this query with my changes works.

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java (hunk 4 above)
Comment (vpapavas) [reply ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:406]: @agavra The approach of putting the "common" code in `process` doesn't work. Reason being is that if the parent expression exists in the group by e.g. `(a+b)`, it will set the flag `foundExpressionInGroupBy` to true and will continue visiting the children, `a` and `b`. If a child doesn't exist in the group by, it will set the flag to false. All the rest of the children then (`b` in this example), will fail and will wrongly get added to the `nonAggParams`. 
TLDR: the flag must be set to false only after the entire tree of the expression has been visited. And that is only possible to do in the specific method for each expression

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java [hunk 9]
```diff
@@ -359,5 +359,46 @@ public Void visitQualifiedColumnReference(
...
+      if (groupBy.contains(node)) {
+        foundExpressionInGroupBy = true;
+      }
+      super.visitSubscriptExpression(node, context);
+      foundExpressionInGroupBy = false;
+      return null;
+    }
+
+    @Override
+    public Void visitArithmeticBinary(final ArithmeticBinaryExpression node, final Void context) {
```
Comment (agavra) [inline ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:374]: **tl;dr** this comment is a little hard to follow because I was understanding the problem as I wrote the comment. basically we need to make sure that only the _first_ node that sets `foundExpressionInGroupBy=true` can set it back to `false`.

---

so here's a freak scenario:
```sql
SELECT as_value(a['idx'] + b['idx']) FROM s GROUP BY a['idx'], a['idx'] + b['idx']
```

visit `a['idx'] + b['idx']` and set `foundExpressionInGroupBy = true`
visit (left)  `a['idx']` and set `foundExpressionInGroupBy = true`
visit `'idx'`
go back to `a['idx']` and set `foundExpressionInGroupBy = false`
visit (right) `b['idx']` and fail because `b['idx']` is not in group by

I think there are actually a few bugs here:
1. is what I showed with the weird group by scenario
1. there is another bug where we shouldn't set `foundExpressionInGroupBy` to `false` unless _we_ set it to true

that (number 2 above) could be an even simpler bug:

```sql
-- a is map<string, map<string, string>>
SELECT as_value(a[b][c]) FROM s GROUP BY a[b][c];
```

we set `foundExpressionInGroupBy` to `true` when we visit `a[b][c]` but then set it to false when we visit `a[b]` causing us to fail when we visit `[c]`

I think the solution here is to have a more complicated `foundExpressionInGroupBy` that once it is set, nobody else can set/unset it until we get back to whoever it was that first set it. You can use a tuple of `(foundExpressionInGroupBy, theGroupByExpression)` to achieve this

please add tests for this too!

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java (hunk 4 above)
Comment (agavra) [reply ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:364]: look at my newest comment, I think we have that problem anyway. I still think we should be able to do this in the common process if you take the suggestion I have there.

Comment (agavra) [reply ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:364]: https://github.com/confluentinc/ksql/pull/5998#discussion_r473371482

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java [hunk 10]
```diff
@@ -276,17 +273,31 @@ private void enforceAggregateRules() {
...
 
+  /**
+   * This visitor performs two tasks: Create the input schema to the AggregateNode and validations.
+   *
+   * For creating the input schema, it checks if any expression along the path from the root
+   * expression to the leaf (UnqualifiedColumnReference) is part of the groupBy. If at least one is,
+   * then the UnqualifiedColumnReference is added to the schema.
+   *
+   * For validation, the visitor checks that:
+   * 1) expressions in non-aggregate functions are part of the grouping clause,
```
Comment (agavra) [inline ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:284]:
```suggestion
   * 1) expressions not in aggregate functions are part of the grouping clause,
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java [hunk 11]
```diff
@@ -276,17 +273,31 @@ private void enforceAggregateRules() {
...
+   * 2) aggregate functions are not nested
+   * 3) window clauses (windowstart, windowend) don't appear in aggregate functions or groupBy
+   * 4) aggregate functions don't appear in the groupBy clause
+   * 5) expressions in the having clause are either aggregate functions or grouping keys
+   */
```
Comment (agavra) [inline ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:289]: first, yay for comments! Thanks for writing this detailed comment, helps future people (including me when I forget what this does) understand it 🎉 

second, nit: i'm pretty sure checkstyle forces the following format, but if it doesn't it'll make it easier to read anyway :D 

```suggestion
  /**
   * This visitor performs two tasks: Create the input schema to the AggregateNode and validations.
   *
   * <p> For creating the input schema, it checks if any expression along the path from the root
   * expression to the leaf (UnqualifiedColumnReference) is part of the groupBy. If at least one is,
   * then the UnqualifiedColumnReference is added to the schema.
   *
   * <p>For validation, the visitor checks that:
   * <ol>
   *   <li> expressions in non-aggregate functions are part of the grouping clause </li>
   *   <li> aggregate functions are not nested </li>
   *   <li> window clauses (windowstart, windowend) don't appear in aggregate functions or groupBy </li>
   *   <li> aggregate functions don't appear in the groupBy clause </li>
   *   <li> expressions in the having clause are either aggregate functions or grouping keys </li>
   * </ol>
   */
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java [hunk 12]
```diff
@@ -300,6 +311,19 @@ private AggregateVisitor(
       this.dereferenceCollector = requireNonNull(dereferenceCollector, "dereferenceCollector");
     }
 
+    @Override
+    public Void process(final Expression node, final Void context) {
+      if (groupBy.contains(node) && currentlyInExpressionPartOfGroupBy == null) {
+        currentlyInExpressionPartOfGroupBy = new Pair<>(node, true);
```
Comment (agavra) [inline ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:317]: do we ever actually use the boolean? it looks like we proxy for it by checking `!= null`

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json [hunk 13]
```diff
@@ -514,6 +514,22 @@
...
+        "CREATE TABLE OUTPUT AS SELECT col1['foo']+col2['bar'] AS G1, AS_VALUE(col1['foo']+col2['bar']), COUNT(*) FROM input GROUP BY col1['foo']+col2['bar'];"
+      ],
+      "inputs": [
+        {"topic": "test_topic", "key": 0, "value": {"col1": {"a" : 1}, "col2": {"b" : 1}}},
+        {"topic": "test_topic", "key": 1, "value": {"col1": {"foo" : 1}, "col2": {"bar" : 1}}},
+        {"topic": "test_topic", "key": 2, "value": {"col1": {"bar" : 1}, "col2": {"foo" : 1}}},
+        {"topic": "test_topic", "key": 3, "value": {"col1": {"foo" : 1}, "col2": {"foo" : 1}}}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "key": 2, "value": {"KSQL_COL_0": 2, "KSQL_COL_1": 1}}
```
Comment (agavra) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json:530]: shouldn't there be four outputs here? or am i missing something?

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java (hunk 10 above)
Comment (vpapavas) [reply ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:284]: It's actually correct what I wrote :P  It will throw if expressions used in aggregate functions are not part of the group by

Comment (agavra) [reply ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:284]: >  It will throw if expressions used in aggregate functions are not part of the group by

now I'm more confused 😂 expressions in aggregate functions must not be part of the group by - `SELECT id, SUM(foo) FROM s GROUP BY id` works (foo is an aggregate function parameter not in group by)

What you had originally written is true, but does not cover everything. It says expressions in UDFs (not UDAFs) must be part of the grouping clause (true). But it is also true that even if they're not in a UDF, as long as they're not in an aggregate function, they must be in the grouping clause:

```sql
--- valid
SELECT non_agg_fun(foo), COUNT(*) from s GROUP BY foo;
SELECT foo, COUNT(*) from s GROUP BY foo;

-- invalid
SELECT foo, COUN(T*) from s GROUP BY bar;
```

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json (hunk 13 above)
Comment (vpapavas) [reply ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json:530]: There is only one input record whose maps have keys `foo` and `bar` respectively

Comment (agavra) [reply ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json:530]: ah! got it - I missed that, thanks for the clarification

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java (hunk 10 above)
Comment (vpapavas) [reply ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:284]: Ah sorry, I meant expressions in non-aggregate functions must be part of group-by. So much confusion :) So, the comments say that the validator will throw if any of the conditions does not hold

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java (hunk 12 above)
Comment (vpapavas) [reply ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:317]: You are right, will change it. The initial intention was to use it

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java (hunk 10 above)
Comment (vpapavas) [reply ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java:284]: Ah wait, I think I know get what you are suggesting. You are suggesting to make the sentence more general. OK, will make the change
